src/err/err_state.rs

```diff
@@ -69,25 +79,47 @@ impl PyErrState {
         // This process is safe because:
         // - Access is guaranteed not to be concurrent thanks to `Python` GIL token
         // - Write happens only once, and then never will change again.
-        // - State is set to None during the normalization process, so that a second
-        //   concurrent normalization attempt will panic before changing anything.
 
-        // FIXME: this needs to be rewritten to deal with free-threaded Python
-        // see https://github.com/PyO3/pyo3/issues/4584
+        // Guard against re-entrant normalization, because `Once` does not provide
+        // re-entrancy guarantees.
+        if let Some(thread) = self.normalizing_thread.lock().unwrap().as_ref() {
+            assert!(
+                !(*thread == std::thread::current().id()),
+                "Re-entrant normalization of PyErrState detected"
+            );
+        }
 
-        let state = unsafe {
-            (*self.inner.get())
-                .take()
-                .expect("Cannot normalize a PyErr while already normalizing it.")
-        };
+        // avoid deadlock of `.call_once` with the GIL
+        py.allow_threads(|| {
```

Comment:
I guess somehow dropping the GIL somehow allows a race condition to happen where multiple threads try to simultaneously create a module...

Reply:
Yeah, I think it's a combination with `GILOnceCell` in the `test_declarative_module`; we allow racing in `GILOnceCell` under the condition where switching the GIL, so this module does actually attempt to get created multiple times. I think it's a bug in `GILOnceCell`, but this also just makes me dislike this lazy stuff even more...

Reply:
I guess this is just a fundamental issue with `GILOnceCell` being racey if the code it wraps ever drops the GIL.

Reply:
I've opened #4676, if I apply that patch on this branch, the problem goes away.